### ISSUE TITLE
[QIR] String interpolation with array results in use after free

### ISSUE THREAD

**Issue by swernli** [microsoft/qsharp-compiler#974]
When displaying an array via string interpolation, the generated QIR includes a loop over string concatenation to assemble each of the items into a single string, such that `[1, 2]` becomes the string `"[1, 2]"`. However, the code incorrectly includes the addition of the opening `"["` inside the body of the loop instead of beforehand, and then tries to decrement the reference count on each iteration through the loop. So this works:
```qsharp
Message($"{[1]}");
```
because the loop over string concatenation only runs once, but this will crash:
```qsharp
Message($"{[1, 2]}");
```
Looking at the QIR directly, I can see where there are a few lines that should be moved up and out of the loop body:
```llvm
  ; THIS IS "["
  %21 = call %String* @__quantum__rt__string_create(i8* getelementptr inbounds ([2 x i8], [2 x i8]* @3, i32 0, i32 0))
  ; THIS IS ", "
  %22 = call %String* @__quantum__rt__string_create(i8* getelementptr inbounds ([2 x i8], [2 x i8]* @4, i32 0, i32 0))
  ; THIS IS "]"
  %23 = call %String* @__quantum__rt__string_create(i8* getelementptr inbounds ([2 x i8], [2 x i8]* @5, i32 0, i32 0))
  br label %header__2

header__2:                                        ; preds = %exiting__2, %exit__1
  %24 = phi i64 [ 0, %exit__1 ], [ %32, %exiting__2 ]
  %25 = icmp sle i64 %24, 1
  br i1 %25, label %body__2, label %exit__2

body__2:                                          ; preds = %header__2
  %26 = call i8* @__quantum__rt__array_get_element_ptr_1d(%Array* %16, i64 %24)
  %27 = bitcast i8* %26 to i64*
  %28 = load i64, i64* %27, align 4
  %29 = call %String* @__quantum__rt__int_to_string(i64 %28)
  ; "[" gets concatenated in loop body when it should be before it
  %30 = call %String* @__quantum__rt__string_concatenate(%String* %21, %String* %29)
  ; Ref count for "[" is decremented in each loop, so the next loop %21 is already freed
  call void @__quantum__rt__string_update_reference_count(%String* %21, i32 -1)
  call void @__quantum__rt__string_update_reference_count(%String* %29, i32 -1)
  %31 = call %String* @__quantum__rt__string_concatenate(%String* %30, %String* %22)
  call void @__quantum__rt__string_update_reference_count(%String* %30, i32 -1)
  br label %exiting__2

```

**Comment by swernli** [comment]
One of the side effects of this issue is that the generated QIR for the entrypoint variant that displays the results via `Message` will not work if the return value of the Q# entrypoint is an array. For example, if the Q# is:
```qsharp
    @EntryPoint()
    operation EntryPoint() : Int[] {
        return [1,2,3];
    }
```
Then running the resulting QIR function `void @Microsoft__Quantum__Qir__Emission__EntryPoint() #1` will crash.

**Comment by bettinaheim** [comment]
Reopening due to https://github.com/microsoft/qsharp-compiler/pull/989.

**Comment by bettinaheim** [comment]
Fixed with #990.
